health/health.d/postfix.conf

```diff
@@ -0,0 +1,8 @@
+template: postfix_queue_emails
+on: postfix.qemails
+calc: $emails
+every: 10s
+warn: $this > 100
```

Comment:
Hello @steffenweber ,

Firstly thank you for your suggestion!

I am more than 10 years without to work with `postfix`, but I remember that during the years I used it as my mail server, it was normal to  have more than `100` emails in the queue, because the organization had more `10000` people, and those years email was the main communication method for us.  I have concerns that this warning value is small.

Best regards!

Reply:
I agree that a good value heavily depends on the use-case and size of the organization. It would probably be better to calculate some kind of failure rate but the total number of emails delivered is unknown to Netdata right?

Reply:
Yeah, we currently have no way to see failure rates for postfix.

Even better than that though IMO would be looking at how long the oldes message in the queue has been in the queue (since even for busy sites, messages should not be queued for more than at most a few minutes), but we have no way to watch that either.

Reply:
Do you think it would be useful to have such a default alarm with a very high warning threshold (1000?) such that it can be easily customized? Or do you think it would be better not to have such a default alarm at all?

Reply:
Do we know what that queue is? [I see there is a lot of queues](http://www.postfix.org/QSHAPE_README.html).

@steffenweber can you share `postqueue -p`?

Reply:
I would argue that having a bad alarm is better than having no alarm, but we need to favor false negatives over false positives for our default alarms (IOW, it’s OK if it does not trigger in some cases when it ideally should, but it’s not OK if it triggers regularly in cases when there is no problem).

However, we can mitigate false positives by making it look at the average value over the past few minutes instead of only the current value (so a single spike won’t trip it, but a continuously high value will), and can help reduce the impact of false positives by adding hysteresis to the trip points.

So, instead of the `calc` line, use `lookup: average -10min unaligned of emails` (IOW, take the average of the past 10 minutes), and instead of the current `warn` and `crit` lines,, do something like:

```
warn: $this > (($status >= $WARNING)  ? (900) : (1000))
crit: $this > (($status >= $CRITICAL) ? (9000) : (10000))
```

We probably also want a `delay` line as well, though I’m less sure what that should be.

I think 1k and 10k probably make sense as trip points in most cases, Having more than 1k emails queued simultaneously for more than a few minutes is atypical for a majority of use cases, and 10k is getting into the territory of potential performance issues.

Reply:
@ilyam8 IIRC, the metric is counting total emails reported by `postqueue -p` without differentiating type.

It would be _great_ though if we could update the collector to differentiate between the different queue states, because that would actually give us much more useful info to alert on.

Reply:
I think `postqueue -p` could leak too much personal data and I'm not sure whether it would help a lot. But the output of `qshape.pl deferred` looks interesting:

```
                  T  5 10 20 40 80 160 320 640 1280 1280+
           TOTAL 10  0  0  0  1  0   0   0   0    0     9
       gmail.com  6  0  0  0  0  0   0   0   0    0     6
     example.org  1  0  0  0  0  0   0   0   0    0     1
     example.net  1  0  0  0  0  0   0   0   0    0     1
     example.com  1  0  0  0  1  0   0   0   0    0     0
      example.de  1  0  0  0  0  0   0   0   0    0     1
```

It seems like the last number in the "TOTAL" row contains the number of emails that have been in the deferred queue for more than 1280 minutes (i.e. more than 21 hours).

On the other hand, being notified about a problem with a delay of 21 hours is maybe not that useful after all.

Maybe instead of a default alarm the documentation page could be complemented with an example alarm explicitly mentioning that the thresholds need to be adjusted depending on the use-case?

Reply:
Yeah, deferred emails are what we really want to care about in this case, but there’s not a good way right now for us to get that unless we update the collector.